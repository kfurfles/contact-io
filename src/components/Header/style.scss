.topbar {
    backface-visibility: hidden;
    background-color: white;
    box-shadow: 0 4px 10px rgba(0, 0, 0, .07);
    left: 0;
    padding: 10px 16px;
    position: fixed;
    top: 0;
    width: 100%;
    height: 60px;

    &__logo {
        // display: inline-block;
        // height: 40px;
        // position: absolute;
        // top: 0;
        // left: 50%;
        // transform: translateX(-50%);

        svg {
            height: 100%;
            width: calc(100% - 100px);

            .st0{
                fill: #2C3341;
                stroke: #2C3341;
            }
            .st1{
                fill: transparent;
                stroke:#4C84FF;
            }

            &.completed{
                .st0{fill:#2C3341;}
	            .st1{fill:#4C84FF;}
            }
        }


        position: fixed;
        top: 50%;
        left: 50%;
        background: #fff;
        width: 100%;
        height: 100%;
        transform: translate(-50%, -50%);
        display: flex;

        svg{
            .st0{
                opacity: 0;
            }
        }

    }

    .container {
        display: flex;
        justify-content: center;
        align-items: center;
        position: relative;
    }
}
