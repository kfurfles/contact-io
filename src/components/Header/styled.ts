import styled from '@emotion/styled';

export const Container = styled.header`
    backface-visibility: hidden;
    background-color: white;
    box-shadow: 0 4px 10px rgba(0, 0, 0, .07);
    left: 0;
    padding: 10px 16px;
    position: fixed;
    top: 0;
    width: 100%;
    height: 60px;

    .logo-link {
        position: fixed;
        top: 50%;
        left: 50%;
        background: #fff;
        width: 100%;
        height: 100%;
        transform: translate(-50%, -50%);
        display: flex;

        svg {
            height: 100%;
            width: calc(100% - 100px);

            .st0{
                fill: #2C3341;
                stroke: #2C3341;
                opacity: 0;
            }
            .st1{
                fill: transparent;
                stroke:#4C84FF;
            }

            &.completed{
                .st0{fill:#2C3341;}
	            .st1{fill:#4C84FF;}
            }
        }
    }

    .container {
        display: flex;
        justify-content: center;
        align-items: center;
        position: relative;
    }
`