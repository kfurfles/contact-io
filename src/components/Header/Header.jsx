import React, {useEffect, useRef} from 'react'

import { ReactComponent as LogoSvg } from "./../../assets/img/logo.svg";
import './style.scss'
import anime from "animejs";


const Header = () => {

    const logoEl = useRef(null)

    useEffect(() => {

        const el = logoEl.current.querySelector('path')
        if(!el) return;
        const {width: mainLogoWidth} = logoEl.current.querySelector('path').getBBox()
        const {width: fullLogoWidth} = logoEl.current.getBBox()


        anime.timeline({
            targets: '.contact-logo .st1',
        })
        .add({
            strokeDashoffset: [anime.setDashoffset, 0],
            easing: 'easeInOutSine',
            duration: 1500,
            delay: function (el, i) {
                return i * 250
            },
        })
        .add({
            fill: ['rgba(0,0,0,0)', '#4C84FF'],
            easing: 'easeInOutSine',
            duration: 1000
        })

        Promise.all([
            anime({
                targets: '.contact-logo',
                translateX: [parseInt(fullLogoWidth - mainLogoWidth), 50],
                duration: 1000,
                easing: 'linear',
                delay: '2000'
            }).finished,
            anime({
                targets: '.contact-logo .st0',
                opacity: 1,
                fill: ['rgba(0,0,0,0)', '#2C3341'],
                duration: 1000,
                easing: 'linear',
                delay: '2000'
            }).finished
        ]).then(() => {
            anime.timeline({
                targets: '.topbar'
            })
                .add({
                    height: '100vh'
                })
                .add({
                    targets: '.topbar__logo',
                    opacity: 0,
                    easing: 'linear',
                    duration: 500,
                    endDelay: 1000
                }).finished.then(() => {
                const link = logoEl.current.parentNode
                link.setAttribute('style', `
                    position: static;
                    top: initial;
                    left: initial;
                    background: transparent;
                    width: initial;
                    height: initial;
                    transform: none;
                    display: block;`
                )
                return true
            })
                .then(() => {
                    anime({
                        targets: '.topbar',
                        height: '70px'
                    })
                })

        })

    },[logoEl])

    return (
        <header className="topbar">
            <div className="container">
                <a href="/" className="topbar__logo" >
                <LogoSvg alt="Logo Contact" ref={logoEl} className="contact-logo" />
                </a>
            </div>
        </header>
    )
}

export default Header