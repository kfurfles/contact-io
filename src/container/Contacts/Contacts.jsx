import React, {Fragment, useCallback, useEffect, useState} from 'react'

import './style.scss'
import FilterBar from '../../components/FilterBar'
import Contact from '../../components/Contact'
import {getUsers} from "../../services/api";

const Contacts = () => {

    const [users, setUsers] = useState([])
    const [filterFn, setFilterFn] = useState(() => () => true)

    async function fetchUsers(){
        return await getUsers()
    }

    const filterUsers = useCallback(({ value, key }) =>{

        const filteredFn = !value
            ? () => () => true
            : () => user => user[key].toString().toLowerCase().includes(value.toString().toLowerCase())

        setFilterFn(filteredFn)
        // eslint-disable-next-line react-hooks/exhaustive-deps
    },[filterFn])

    const sortUsers = useCallback(({ direction, key }) =>{
        const sortedUsers = users.sort((a,b) => a[key] > b[key] && direction === 'up' ? 1 : -1)

        setUsers([ ...sortedUsers ])
    },[users])

    useEffect(() => {
        fetchUsers()
            .then(users => {
                setUsers(users)
            })
    },[])

    return (
        <Fragment>
            <div className="container">
                <FilterBar
                    onChangeSort={sortUsers}
                    onChangeFilter={filterUsers}/>
            </div>

            <div className="container">
                <section className="contacts">
                    <article className="contact">
                    <span className="contact__avatar" />
                    <span className="contact__data">Nome</span>
                    <span className="contact__data">Telefone</span>
                    <span className="contact__data">País</span>
                    <span className="contact__data">Admissão</span>
                    <span className="contact__data">Empresa</span>
                    <span className="contact__data">Departamento</span>
                    </article>

                    {users
                        .filter(filterFn)
                        .map(user => <Contact user={user} key={user.id}/> )}

                </section>
            </div>
        </Fragment>
    )
}

export default Contacts