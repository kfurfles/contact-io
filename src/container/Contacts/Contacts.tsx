import React, {Fragment, useCallback, useEffect, useState} from 'react'

// import './style.scss'
import { Container, Contacts as ContactsSection, ContactsHeader } from './styled'
import FilterBar from '../../components/FilterBar'
import Contact from '../../components/Contact'
import { getUsers } from "../../services/api";
import { IUser } from '../../interfaces/IUser'

const Contacts = () => {

    const [users, setUsers] = useState<IUser[]>([])
    const [filterFn, setFilterFn] = useState<() => (user: IUser) => boolean>(() => () => true)

    const fetchUsers = () => getUsers()

    const filterUsers = useCallback(({ value, key }: { value: string, key: keyof IUser }) =>{

        const filteredFn = !value
            ? () => () => true
            : () => (user: IUser) => user['name'].toString().toLowerCase().includes(value.toString().toLowerCase())

        setFilterFn(filteredFn)
        // eslint-disable-next-line react-hooks/exhaustive-deps
    },[filterFn])

    const sortUsers = useCallback(({ direction, key }: { direction: 'up' | 'down', key: keyof IUser }) =>{
        const sortedUsers = users.sort((a,b) => a[key] > b[key] && direction === 'up' ? 1 : -1)

        setUsers([ ...sortedUsers ])
    },[users])

    useEffect(() => {
        fetchUsers()
            .then(users => {
                setUsers(users)
            })
    },[])

    return (
        <Fragment>
            <Container>
                <FilterBar
                    onChangeSort={sortUsers}
                    onChangeFilter={filterUsers}/>
            </Container>

            <Container>
                <ContactsSection>
                    <ContactsHeader>
                        <span className="avatar" />
                        <span className="data">Nome</span>
                        <span className="data">Telefone</span>
                        <span className="data">País</span>
                        <span className="data">Admissão</span>
                        <span className="data">Empresa</span>
                        <span className="data">Departamento</span>
                    </ContactsHeader>

                    {users
                        .filter(filterFn)
                        .map(user => <Contact user={user} key={user.id}/> )}

                </ContactsSection>
            </Container>
        </Fragment>
    )
}

export default Contacts