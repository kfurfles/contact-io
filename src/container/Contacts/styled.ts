import styled from '@emotion/styled'

export const Container = styled.div`
    width: 100%;
    margin: 0 auto;
    padding: 0 15px;

    @media screen and (min-width: 960px) {
        width: 960px;
    }
`

export const Contacts = styled.section`
    width: 100%;
    display: flex;
    flex-direction: column;
    padding: 16px;

    @media screen and (min-width: 728px) {
        padding: 16px 0;
    }
`

export const ContactsHeader = styled.div`
    display: flex;
    justify-content: space-between;
    flex-wrap: nowrap;
    font-size: 1.2rem;
    margin-bottom: 20px;
    overflow: hidden;
    padding: 5px;

    &:first-of-type {
        margin-bottom: 5px;
        background-color: transparent;
        border: none;
        box-shadow: none;
        padding: 0 5px 0 50px;
        font-weight: bold;

        & .avatar {
            height: 0;
            border: none;
        }

        .data:nth-of-type(2) {
            color: #000;
        }
    }

    .data{
        width: calc(100% / 6);
        padding: 0 10px;
    }
`
