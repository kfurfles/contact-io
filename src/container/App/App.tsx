import React from 'react';

import './App.scss';
import Contacts from '../Contacts';
import Header from '../../components/Header';

const App = () => {

  return (
  <React.Fragment>
    <Header />
    <Contacts />
  </React.Fragment>)
}

export default App;
