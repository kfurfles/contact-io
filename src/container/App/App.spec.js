import React from 'react';
import { render } from '@testing-library/react';

import App from './index';
import Header from "../../components/Header";
import Contacts from "../Contacts";


describe('App', () => {
    it('Should render App', () => {
        const { fragment } = render(
            <App>
                <Header />
                <Contacts />
            </App>);
        expect(fragment).toMatchSnapshot();
    });

});