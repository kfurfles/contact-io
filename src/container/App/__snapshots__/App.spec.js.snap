// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should render App 1`] = `undefined`;
